### ISSUE TITLE
Faster Directory Listings in VFS

### ISSUE THREAD

**Issue by darthShadow** [rclone/rclone#6632]
#### The associated forum post URL from `https://forum.rclone.org`

N/A

#### What is your current rclone version (output from `rclone version`)?

N/A (but `v1.61.0-beta.6609.ca23c6435`)

#### What problem are you are trying to solve?

Slow directory listings in VFS when backed by a Ceph mount.

#### How do you think rclone should be changed to solve that?

Generally, the Ceph mount is fast enough that these issues are negligible but due to unavoidable issues, the Ceph mount has become slower temporarily which has highlighted these issues.

There are 2 approaches that I can see:

* For the Local Ceph FS: Do the listing of the directory in parallel?
* For the VFS: When the directory is expired, perhaps it can be loaded lazily i.e. return the stale entries till the remote has finished listing the directory and then replace the entries? 

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by darthShadow** [comment]
I am mostly looking for thoughts on the above approaches or any alternatives so I can experiment with those and come back with proper numbers and maybe even a PR.

**Comment by darthShadow** [comment]
> Which CEPH interface are you using? S3 or Swift?

This is a CephFS kernel mount so effectively a local filesystem for all intents and purposes. Perhaps we could parallelize at least the `os.Stat` calls?

> but it effectively breaks the directory caching scheme.

Yeah, I have been unable to think of a good way to do this.

The problem is sometimes the MDS servers or remotes can take time (especially with larger directory trees) and any further listing commands on that will get stuck waiting on the locks to be released.

**Comment by darthShadow** [comment]
> Do the directories have many entries?

Yeah, some directories have upto 40k entries.

**Comment by ncw** [comment]
> > Which CEPH interface are you using? S3 or Swift?
> 
> This is a CephFS kernel mount so effectively a local filesystem for all intents and purposes. Perhaps we could parallelize at least the `os.Stat` calls?

That is very do-able and a great idea.

It might also be worth revisiting the listing interface for the local backend. At the moment unix does a file listing then stats each file, whereas windows gets the listing with stats straight from the kernel.

If there was an interface like that for linux it would speed stuff up enormously.

I have a feeling the go internal interfaces got re-organised 

It would be worth investigating how https://pkg.go.dev/os#File.ReadDir works and seeing if it uses the low level interface to read directory names + stat from the kernel or whether it just does read file names then stat each one.

> > but it effectively breaks the directory caching scheme.
> 
> Yeah, I have been unable to think of a good way to do this.
> 
> The problem is sometimes the MDS servers or remotes can take time (especially with larger directory trees) and any further listing commands on that will get stuck waiting on the locks to be released.

That is the problem with the VFS in general and file systems. They are expected to be 100% reliable and when things go wrong it doesn't cope well (unlike rclone sync).

**Comment by darthShadow** [comment]
> I have a feeling the go internal interfaces got re-organised

Yep, was looking into this yesterday and found this issue: https://github.com/golang/go/issues/16399

It doesn't help us directly since we need the stat info too but gives 2 promising libraries that I can compare against:
  * https://github.com/karrick/godirwalk
  * https://github.com/saracen/walker

Will try them out and see how well they work.

---

The locks have been particularly problematic. I have tried comparing the rclone mount on top of the Ceph mount vs. just the Ceph mount and I have observed that listing the same directory in the Ceph mount from multiple sessions/windows don't block on each other unlike the rclone mount which blocks on further directory listings till the first listing has returned (all after the directory cache has expired).

I am not sure if doing something similar to this would be easier or even feasible?

**Comment by ncw** [comment]
> Yep, was looking into this yesterday and found this issue: [golang/go#16399](https://github.com/golang/go/issues/16399)
> 
> It doesn't help us directly since we need the stat info too but gives 2 promising libraries that I can compare against:
> 
> * https://github.com/karrick/godirwalk
> * https://github.com/saracen/walker

I looked at the code for those and they don't use any special kernel interface as far as I can see.

So I think the best idea for a speedup would be to do the stat's in parallel in the existing code. That will mean we can control stat vs lstat and the error handling exactly how we need it.

> The locks have been particularly problematic. I have tried comparing the rclone mount on top of the Ceph mount vs. just the Ceph mount and I have observed that listing the same directory in the Ceph mount from multiple sessions/windows don't block on each other unlike the rclone mount which blocks on further directory listings till the first listing has returned (all after the directory cache has expired).
> 
> I am not sure if doing something similar to this would be easier or even feasible?

I think the CEPH mount is a full kernel filing system isn't it, not fuse? If so it has a significant advantage.

If we don't mind returning stale data from directory listings then we can use a two phase update - return the old data until the new data is ready. This will affect consistency but maybe not very much as rclone's virtual directory entries deal with locally created files.

**Comment by darthShadow** [comment]
> That will mean we can control stat vs lstat and the error handling exactly how we need it.

Sounds good. Will give it a try and come back with the results.

---

> I think the CEPH mount is a full kernel filing system isn't it, not fuse?

Yep, I just thought the idea was interesting. We could let multiple listings proceed in parallel and only lock once any of those return (and consequently cancel the others). This is simple (in terms of transaction costs) enough on local filesystems since there is no price to pay there ... 😉 

> returning stale data from directory listings

Yep, this is exactly what I meant by loading it lazily in the first post. Just got confused when you mentioned that it would break the directory caching and figured I was missing something in the implementation that wouldn't be handled by the virtual entries ... 😅 

Any thoughts on how to to go about trying this approach? Specifically what the locking etc. should look like?

**Comment by ncw** [comment]
Let's try the parallel stat first.

Changing the locking for the directory reads is very complicated! The locks are mostly about protecting the directory map so changing that structure would possibly be necessary. Or maybe having two, a current one and one being updated.

I'm worried about the stale listing though. It could potentially always be one listing behind.

**Comment by darthShadow** [comment]
Sounds good. Will come back when I have some numbers.

**Comment by Skylion007** [comment]
> > Yep, was looking into this yesterday and found this issue: [golang/go#16399](https://github.com/golang/go/issues/16399)
> > It doesn't help us directly since we need the stat info too but gives 2 promising libraries that I can compare against:
> > 
> > * https://github.com/karrick/godirwalk
> > * https://github.com/saracen/walker
> 
> I looked at the code for those and they don't use any special kernel interface as far as I can see.
> 
> So I think the best idea for a speedup would be to do the stat's in parallel in the existing code. That will mean we can control stat vs lstat and the error handling exactly how we need it.
> 
> > The locks have been particularly problematic. I have tried comparing the rclone mount on top of the Ceph mount vs. just the Ceph mount and I have observed that listing the same directory in the Ceph mount from multiple sessions/windows don't block on each other unlike the rclone mount which blocks on further directory listings till the first listing has returned (all after the directory cache has expired).
> > I am not sure if doing something similar to this would be easier or even feasible?
> 
> I think the CEPH mount is a full kernel filing system isn't it, not fuse? If so it has a significant advantage.
> 
> If we don't mind returning stale data from directory listings then we can use a two phase update - return the old data until the new data is ready. This will affect consistency but maybe not very much as rclone's virtual directory entries deal with locally created files.

I am no golang expert by any means, but Isn't the way to fix this issue just to use the recommend filepath.WalkDir now that we are on a recent version golang? https://github.com/golang/go/issues/16399#issuecomment-776327385 . Seems like the purpose of filepath.WalkDir is explicitly to avoid os.stat calls in a similar way to how GNU find works. https://pkg.go.dev/path/filepath@master#WalkDir I am pretty sure this uses the same kernel interface as Python's os.scandir and I have noticed massive speedups in my Python scripts when iterating over NFS mounts. Or is the issue that when the filepath.WalkDir cannot make a fused kernel call, that it defaults to a slow non parallel implementation?

**Comment by Skylion007** [comment]
@ncw Here is a blogpost that explicitly states that the main reason for adding ReadDIr is performance so it's definitely better when it's possible to use it: https://benhoyt.com/writings/go-readdir/

**Comment by ncw** [comment]
Interesting blog post @Skylion007 - thanks.

I don't think using this would be a big win unless we decide to delay doing the stat on the file until it is actually needed. Using ReadDir we can tell files from directories so we could create the correct kind of local object without doing a stat, and do the stat when some info is needed from the object. There are complications here with symlinks so we'd probably need to stat them immediately.

This is certainly a possibility and it might help with @darthShadow 's problem.

**Comment by Skylion007** [comment]
> Interesting blog post @Skylion007 - thanks.
> 
> I don't think using this would be a big win unless we decide to delay doing the stat on the file until it is actually needed. Using ReadDir we can tell files from directories so we could create the correct kind of local object without doing a stat, and do the stat when some info is needed from the object. There are complications here with symlinks so we'd probably need to stat them immediately.
> 
> This is certainly a possibility and it might help with @darthShadow 's problem.

@ncw One additional point of clarification not mentioned in the blogpost is that using io.ReadDir() can actually fuse the stat calls during the directory listing if it's supported by the OS and filesystem and cache it. For instance, NFSv3 has the [ReadDirPlus](https://www.freesoft.org/CIE/RFC/1813/37.htm) command that can be sent to the NFS server and reads a directory on the server AND get additional info from the inodes at the same time (for instance file permision / size of the file / time of modification etc). NFSV4 has this functionality too but it's merged into the normal ReadDir call as arguments).

This FileInfo object is then cached in the fs.DirEntry object for subsequent stat calls. If the backend does not support returning all this info in a fused call, than the stat() call is performed lazily (and may be cached).

This is hinted in the golang docs, but it really should be explicitly stated:
```
        // Info returns the FileInfo for the file or subdirectory described by the entry.
        // The returned FileInfo may be from the time of the original directory read
	// or from the time of the call to Info.
```
This is the same way it's implemented in Python by the way At worse, it's the same as calling stat on each file, at best it's a amortized constant time operation since the info is returned by the OS at directory traversal time.

**Comment by ncw** [comment]
@Skylion007 does it actually fuse the request on Linux? I looked at the code to see if I could work out if it did and I didn't think so, but I may be wrong! I believe it does on windows though (but rclone does that already).

**Comment by darthShadow** [comment]
Sorry for having been silent but this has dropped in my priority list significantly since the recent updates to bazil/fuse which enabled parallel dirops:

https://github.com/bazil/fuse/commit/a93b180d7a2e1b0944b236f5295d4747c9808a73

https://libfuse.github.io/doxygen/fuse-3_813_80_2include_2fuse__common_8h.html#a3311a7f569e9fbf55b3f0bcaadbe4ad0

This has significantly reduced the occasional lockups due to slow MDS servers.

I don't think adding parallel stat calls would provide any significant benefit upon this. The other major benefit would be the lazy directory cache invalidation but since that would be a lot more complex, I haven't given this much thought recently.

**Comment by ncw** [comment]
> Sorry for having been silent but this has dropped in my priority list significantly since the recent updates to bazil/fuse which enabled parallel dirops:
> 
> [bazil/fuse@a93b180](https://github.com/bazil/fuse/commit/a93b180d7a2e1b0944b236f5295d4747c9808a73)
> 
> https://libfuse.github.io/doxygen/fuse-3_813_80_2include_2fuse__common_8h.html#a3311a7f569e9fbf55b3f0bcaadbe4ad0
> 
> This has significantly reduced the occasional lockups due to slow MDS servers.

Nice :-)

> I don't think adding parallel stat calls would provide any significant benefit upon this. The other major benefit would be the lazy directory cache invalidation but since that would be a lot more complex, I haven't given this much thought recently.

:+1:

**Comment by darthShadow** [comment]
It did improve the directory listing performance but at the cost of a sharp rise in the CPU usage due to extensive GC:

![Screenshot 2023-03-30 at 11 58 41 AM](https://user-images.githubusercontent.com/6824881/228800942-ad9e4584-6dfc-4d41-90cd-d222480106bf.jpg)

Let me see if I can improve some low-hanging fruits by using sync.Pool.

**Comment by ncw** [comment]
Note that it is my plan to replace the in memory directory listings with an on disk database based off https://pkg.go.dev/github.com/rclone/rclone/lib/kv which I think should help a lot with all sorts of VFS issues.

I have done quite a bit of work towards this but it won't be ready for a while yet!

**Comment by ncw** [comment]
> Good improvements so far with ~50% reduction in CPU Usage:

Great work :-)

> I know go-fuse has proven difficult to integrate in the past but is there anyway you could take another attempt at it? bazil/fuse seems to be stuck in a state of limbo again and cgofuse has also not added fuse3 support yet.

We already integrated go-fuse as `rclone mount2` - give it a go.

**Comment by darthShadow** [comment]
> We already integrated go-fuse as rclone mount2 - give it a go.

Ah, so it is feature complete, just untested. I was aware of the alternative but wasn't sure if any functionality was missing or just untested.

**Comment by ncw** [comment]
@darthShadow wrote:
> > We already integrated go-fuse as rclone mount2 - give it a go.
> 
> Ah, so it is feature complete, just untested. I was aware of the alternative but wasn't sure if any functionality was missing or just untested.

It passes the integration tests too so it is tested in some sense. It just hasn't received the wider testing that `rclone mount` has!

**Comment by darthShadow** [comment]
Started testing with mount2 a day ago and seems fine so far.

Required 2 commits:
* Updating the go-fuse version, since there's no recent tag: https://github.com/darthShadow/rclone/commit/796306b38c42c535f6a2737221392759d2878e55
* Fixing the statsfs API: https://github.com/darthShadow/rclone/commit/75b879db9d9ca8ba957f0409446f022e4aedfa60

Do you want me to PR either or both of these changes?

And also the earlier bazil/fuse changes? They are going to require switching to a custom fork for the changes with larger impact which may not be a good idea but a single change can probably be brought in: https://github.com/darthShadow/rclone/commit/eb6567689fa4a1404d55c152069236d2fcc48095#diff-3a08174dbb4f0d8dabc74fee9f004b9090f1b0f69947b267947912a78f84e3b3

**Comment by ncw** [comment]
> Started testing with mount2 a day ago and seems fine so far.

Great
 
> Required 2 commits:
> 
> * Updating the go-fuse version, since there's no recent tag: [darthShadow@796306b](https://github.com/darthShadow/rclone/commit/796306b38c42c535f6a2737221392759d2878e55)
> * Fixing the statsfs API: [darthShadow@75b879d](https://github.com/darthShadow/rclone/commit/75b879db9d9ca8ba957f0409446f022e4aedfa60)
> 
> Do you want me to PR either or both of these changes?

Yes please - they look useful.
 
> And also the earlier bazil/fuse changes? They are going to require switching to a custom fork for the changes with larger impact which may not be a good idea but a single change can probably be brought in: [darthShadow@eb65676#diff-3a08174dbb4f0d8dabc74fee9f004b9090f1b0f69947b267947912a78f84e3b3](https://github.com/darthShadow/rclone/commit/eb6567689fa4a1404d55c152069236d2fcc48095#diff-3a08174dbb4f0d8dabc74fee9f004b9090f1b0f69947b267947912a78f84e3b3)

I think there is a bit more in that commit than there needs to be (the go.mod stuff) but it seems to be removing a memory allocation in `Read` which sounds like a very good idea.

**Comment by darthShadow** [comment]
#6982 & #6983

**Comment by darthShadow** [comment]
Looks like there is Go library too for interacting with Ceph clusters: https://github.com/ceph/go-ceph

This also includes a watcher for any directory/file changes (I hope): https://github.com/ceph/go-ceph/blob/71cf02d67ff4e8381869f4e3cbf910d47b6f7c2a/rados/watcher.go

**Comment by ncw** [comment]
> Looks like there is Go library too for interacting with Ceph clusters: https://github.com/ceph/go-ceph

Interesting. It requires CGO though which is a pain.

Perhaps it could be made into a `ceph` backend.

**Comment by darthShadow** [comment]
Waiting on something like inotify to be available for cache invalidations before I give it a try: https://github.com/ceph/go-ceph/issues/478

**Comment by darthShadow** [comment]
I think we could use kopia's implementation for readdir to process it more efficiently and faster: https://github.com/kopia/kopia/blob/8b0296cdf2c37d4f0f3de408eb653b80b56a23a5/fs/localfs/local_fs.go#L129

**Comment by nh2** [comment]
I've successfully used Kopia for my scenarios from #7153 as well, but the hardcode of `workers := 16` was a limitation on my Ceph cluster. An option would be better.

**Comment by Masamune3210** [comment]
Not to interject, but how would I go about testing mount2? It is refused as an unknown command when trying rclone mount2, I'm assuming I am missing something.

**Comment by ncw** [comment]
I've put this on the 1.64 milestone to remind me to have a go at this as I don't think it is hard. Unless you want to @darthShadow ?

@Masamune3210 `rclone mount2` only works on linux and macOS/amd64 right now.

**Comment by Masamune3210** [comment]
Aah, that makes sense. Thanks

**Comment by darthShadow** [comment]
> Unless you want to @darthShadow ?

Yeah, I was planning to get this done when I get some free time in the next week or two.

**Comment by darthShadow** [comment]
Updated the tags & title to reflect the current approach. The lazy loading approach will be tracked as a separate issue: https://github.com/rclone/rclone/issues/7171

**Comment by darthShadow** [comment]
And the idea for a Ceph Backend: https://github.com/rclone/rclone/issues/7172
